stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/adapter/ChannelListItemAdapter.kt

```diff
@@ -12,32 +11,55 @@ import io.getstream.chat.android.ui.channel.list.adapter.viewholder.ChannelListL
 import io.getstream.chat.android.ui.utils.extensions.cast
 import io.getstream.chat.android.ui.utils.extensions.diff
 import io.getstream.chat.android.ui.utils.extensions.firstOrDefault
+import io.getstream.chat.android.ui.utils.extensions.safeCast
 
 internal class ChannelListItemAdapter(var style: ChannelListViewStyle) :
-    ListAdapter<Channel, BaseChannelListItemViewHolder>(DIFF_CALLBACK) {
+    ListAdapter<ChannelListItem, BaseChannelListItemViewHolder>(DIFF_CALLBACK) {
 
     var viewHolderFactory: ChannelListItemViewHolderFactory = ChannelListItemViewHolderFactory()
 
     val listenerContainer: ChannelListListenerContainer = ChannelListListenerContainer()
 
-    var endReached: Boolean = true
-        set(value) {
-            if (value) {
-                // if we've reached the end, remove the last item
-                notifyItemRemoved(itemCount)
-            }
-            field = value
-        }
-
     companion object {
-        val DIFF_CALLBACK: DiffUtil.ItemCallback<Channel> = object : DiffUtil.ItemCallback<Channel>() {
-            override fun areItemsTheSame(oldItem: Channel, newItem: Channel): Boolean = oldItem.cid == newItem.cid
-
-            override fun areContentsTheSame(oldItem: Channel, newItem: Channel): Boolean =
-                !oldItem.diff(newItem).hasDifference()
-
-            override fun getChangePayload(oldItem: Channel, newItem: Channel): Any = oldItem.diff(newItem)
-        }
+        val DIFF_CALLBACK: DiffUtil.ItemCallback<ChannelListItem> =
+            object : DiffUtil.ItemCallback<ChannelListItem>() {
+                override fun areItemsTheSame(oldItem: ChannelListItem, newItem: ChannelListItem): Boolean {
+                    if (oldItem::class != newItem::class) {
+                        return false
+                    }
+
+                    return when (oldItem) {
+                        is ChannelListItem.ChannelItem -> {
+                            oldItem.channel.cid == newItem.safeCast<ChannelListItem.ChannelItem>()?.channel?.cid
+                        }
+
+                        else -> true
+                    }
+                }
+
+                override fun areContentsTheSame(oldItem: ChannelListItem, newItem: ChannelListItem): Boolean {
+                    // this is only called if areItemsTheSame returns true, so they must be the same class
+                    return when (oldItem) {
+                        is ChannelListItem.ChannelItem -> {
```

Comment:
The observation, you can use only boolean logic here. No need for the `when`

Reply:
The `when` is just for implicit smart casting & readability